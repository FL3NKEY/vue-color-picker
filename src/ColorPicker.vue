<template>
    <div tabindex="0"
         class="rcp"
         :class="{ 'dragging': isDragging }"
         @keyup.enter="selectColor"
         @keydown.up.right.prevent="rotate($event, true)"
         @keydown.down.left.prevent="rotate($event, false)">
        <div class="rcp__palette"
             :class="isPaletteIn ? 'in' : 'out'"
             @transitionend="toggleKnob"
             ref="palette">
            <canvas></canvas>
        </div>

        <div class="rcp__rotator"
             :style="{ 'pointer-events': isDisabled ? 'none' : null }"
             @dblclick.self="rotateToMouse"
             ref="rotator"
             @transitionend="hidePalette">
            <div class="rcp__knob" :class="isKnobIn ? 'in' : 'out'"></div>
        </div>

        <div class="rcp__ripple"
             :class="{ 'rippling': isRippling }"
             :style="{ borderColor: color }"
             @animationend="stopRipple">
        </div>

        <button type="button"
                class="rcp__selector"
                :class="{ 'pressed': isPressed }"
                :style="{ backgroundColor: color }"
                @animationend="togglePicker"
                @click="selectColor">
        </button>
    </div>
</template>

<script>
    import fillColorWheel from '@radial-color-picker/color-wheel';
    import Rotator from '@radial-color-picker/rotator';

    let rotator;

    export default {
        name: 'vue-color-picker',
        props: {
            value: {
                default: () => ({ hue: 0, saturation: 100, luminosity: 50, alpha: 1 }),
            },
            step: {
                default: 2,
            },
            mouseScroll: {
                default: false,
            },
            variant: {
                default: 'collapsible', // collapsible | persistent
            }
        },
        data() {
            return {
                isPaletteIn: true,
                isKnobIn: true,
                isPressed: false,
                isRippling: false,
                isDragging: false,
                isDisabled: false,
            }
        },
        computed: {
            color() {
                const { hue, saturation = 100, luminosity = 50, alpha = 1 } = this.value;

                return `hsla(${hue}, ${saturation}%, ${luminosity}%, ${alpha})`;
            }
        },
        watch: {
            'value.hue': function(newAngle, oldAngle) {
                if (newAngle != oldAngle) {
                    rotator.angle = newAngle;
                }
            },
        },
        mounted() {
            if (this.mouseScroll) {
                this.$refs.rotator.addEventListener('wheel', this.onScroll);
            }

            this.$refs.palette.style.backgroundImage = 'conic-gradient(red, yellow, lime, aqua, blue, magenta, red)';

            if (!this.$refs.palette.style.backgroundImage) {
                fillColorWheel(this.$refs.palette.firstElementChild, this.$el.offsetWidth || 280);
            }

            rotator = new Rotator(this.$refs.rotator, {
                angle: this.value.hue,
                onRotate: this.updateColor,
                onDragStart: () => {
                    this.isDragging = true;
                },
                onDragStop: () => {
                    this.isDragging = false;
                },
            });
        },
        methods: {
            onScroll(ev) {
                if (this.isDisabled)
                    return;

                ev.preventDefault();

                if (ev.deltaY > 0) {
                    rotator.angle += this.step;
                } else {
                    rotator.angle -= this.step;
                }

                this.updateColor(rotator.angle);
            },
            rotate(ev, isIncrementing) {
                if (this.isDisabled)
                    return;

                let multiplier = isIncrementing ? 1 : -1;

                if (ev.ctrlKey) {
                    multiplier *= 6;
                } else if (ev.shiftKey) {
                    multiplier *= 3;
                }

                rotator.angle += this.step * multiplier;
                this.updateColor(rotator.angle);
            },
            updateColor(hue) {
                this.$emit('input', {
                    hue,
                    saturation: this.value.saturation || 100,
                    luminosity: this.value.luminosity || 50,
                    alpha: this.value.alpha || 1,
                });
            },
            rotateToMouse(ev) {
                if (this.isDisabled)
                    return;

                rotator.setAngleFromEvent(ev);
            },
            selectColor() {
                this.isPressed = true;

                if (!this.isDisabled) {
                    this.$emit('select', this.value);
                    this.isRippling = true;
                } else {
                    this.isPaletteIn = true;
                }
            },
            togglePicker() {
                if (this.variant !== 'persistent') {
                    if (this.isDisabled) {
                        this.isKnobIn = true;
                    } else {
                        this.isKnobIn = false;
                    }
                }

                this.isPressed = false;
            },
            hidePalette() {
                if (!this.isDisabled) {
                    this.isPaletteIn = false;
                } else {
                    this.isDisabled = false;
                }
            },
            stopRipple() {
                this.isRippling = false;
            },
            toggleKnob(ev) {
                // 'transitionend' fires for every transitioned property
                if (ev.propertyName === 'transform') {
                    if (this.isDisabled) {
                        this.isKnobIn = true;
                    } else {
                        this.isDisabled = true;
                    }
                }
            },
        },
        beforeDestroy() {
            rotator.destroy();
            rotator = null;
        },
    };
</script>

<style>
    .rcp, .rcp div, .rcp button {
        -webkit-touch-callout: none;
        -webkit-tap-highlight-color: transparent;
        user-select: none;
        box-sizing: border-box;
    }

    .rcp {
        display: block;
        overflow: hidden;
        width: 280px;
        height: 280px;
        position: relative;
        transform: scale(1.001);
        transition: transform 0.15s cubic-bezier(0.68, 0, 0.47, 2);
    }

    .rcp:focus {
        outline: 0;
    }

    .rcp:hover .rcp__knob {
        box-shadow: 0 0 20px rgba(0, 0, 0, 0.19), 0 0 10px rgba(0, 0, 0, 0.24);
    }

    .rcp.dragging {
        transform: scale(1.04);
    }

    .rcp.dragging .rcp__rotator {
        z-index: 1;
    }

    .rcp__palette {
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        width: 100%;
        height: 100%;
        background-size: 100% 100%;
        border-radius: 50%;
        overflow: hidden;
        will-change: transform, opacity;
        transition: transform .5s cubic-bezier(0.35, 0, 0.25, 1),
                    opacity .5s cubic-bezier(0.35, 0, 0.25, 1);
    }

    .rcp__palette::before {
        box-sizing: border-box;
        content: '';
        display: block;
        position: absolute;
        width: 76%;
        height: 76%;
        top: 12%;
        left: 12%;
        background-color: #fff;
        border-radius: 50%;
    }

    .rcp__palette.in {
        transform: scale(1);
        opacity: 1;
    }

    .rcp__palette.out {
        transform: scale(0);
        opacity: 0;
    }

    .rcp__rotator {
        width: 100%;
        height: 100%;
        position: absolute;
    }

    .rcp__knob {
        box-shadow: 0 0 10px rgba(0, 0, 0, 0.12), 0 0 5px rgba(0, 0, 0, 0.16);
        border-radius: 50%;
        position: absolute;
        width: 7%;
        height: 7%;
        top: 2.5%;
        left: 46.5%;
        background-color: #fff;
        transition: transform .4s cubic-bezier(0.35, 0, 0.25, 1);
        outline: 0;
        border-style: none;
    }

    .rcp__knob.in {
        transform: scale(1);
    }

    .rcp__knob.out {
        transform: scale(0);
    }

    .rcp__selector {
        position: absolute;
        width: 25%;
        height: 25%;
        top: 37.5%;
        left: 37.5%;
        padding: 0;
        margin: 0;
        border-radius: 50%;
        background-color: #ff0000;
        outline: 0;
        cursor: pointer;
        transition: transform .7s cubic-bezier(0.35, 0, 0.25, 1);
        will-change: transform;
        overflow: visible;
        border: 6px solid #fff;
        box-shadow: 0 0 0 1px #b2b2b2;
    }

    .rcp__selector::-moz-focus-inner {
      border: 0;
    }

    .rcp__selector:hover {
        box-shadow: 0 0 1px 1px #333;
    }

    .rcp__selector:focus {
        box-shadow: 0 0 1px 2px #b2b2b2;
    }

    .rcp__selector.pressed {
        animation: rcp-beat .4s cubic-bezier(0.35, 0, 0.25, 1) forwards;
    }

    .rcp__ripple {
        width: 20%;
        height: 20%;
        border-radius: 50%;
        border: #ff0000 solid 8px;
        opacity: 0;
        position: absolute;
        top: 40%;
        left: 40%;
        z-index: -1;
    }

    .rcp__ripple.rippling {
        z-index: 0;
        animation: rcp-ripple .5s cubic-bezier(0.35, 0, 0.25, 1) forwards;
    }

    @keyframes rcp-ripple {
        0%   { transform: scale(1); opacity: .3; }
        50%  { opacity: .1; }
        100% {
            opacity: 0;
            border-width: 0;
            transform: scale(3.8);
        }
    }

    @keyframes rcp-beat {
        0%   { transform: scale(1); }
        25%  { transform: scale(0.8); }
        50%  { transform: scale(1); }
        100% { transform: scale(1); }
    }
</style>
