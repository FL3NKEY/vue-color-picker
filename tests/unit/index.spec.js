import { shallowMount } from '@vue/test-utils';
import ColorPicker from '@/src/ColorPicker.vue';

// canvas support in JSDom is weak and throws an error so the dependency is mocked
import fillColorWheel from '@radial-color-picker/color-wheel';
jest.mock('@radial-color-picker/color-wheel');

describe('Init', () => {
    beforeEach(() => {
        fillColorWheel.mockClear();
    });

    it('renders correctly', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        expect(el.element).toMatchSnapshot();
    });

    it('setups an optional event for updating the hue by scrolling', () => {
        const onScrollStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                saturation: 100,
                luminosity: 50,
                alpha: 1,
                mouseScroll: true,
            },
            methods: {
                onScroll: onScrollStub,
            },
        });

        el.find('.rcp__rotator').trigger('wheel');

        expect(onScrollStub).toHaveBeenCalled();
    });

    // jsdom doesn't support layouting so `getComputedStyle(palette).backgroundImage`
    // would never include conic-gradient
    it('setups a fallback to canvas when `conic-gradient` CSS is not supported', () => {
        shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        expect(fillColorWheel).toHaveBeenCalled();
    });
});

describe('Core Methods', () => {
    it('emits an `input` event when updateColor is called', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        el.vm.updateColor(90);

        expect(el.emitted('input')[0]).toEqual([90]);
    });

    it('starts the pressed animation when selectColor is called', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        el.vm.selectColor();

        expect(el.vm.isPressed).toBe(true);
    });

    it('emits a `select` event when selectColor is called and the picker is opened', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
            },
            data() {
                return {
                    isPaletteIn: true,
                    isKnobIn: true,
                };
            },
        });

        el.vm.selectColor();

        expect(el.emitted('select')[0]).toEqual([30]);
    });

    it('shows the palette when selectColor is called and the picker is closed', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isPaletteIn: false,
                    isKnobIn: false,
                };
            },
        });

        el.vm.selectColor();

        expect(el.vm.isPaletteIn).toBe(true);
    });

    it('closes the picker when togglePicker is called and the picker is opened', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isKnobIn: true,
                };
            },
        });

        el.vm.togglePicker();

        expect(el.vm.isKnobIn).toBe(false);
    });

    it('opens the picker when togglePicker is called and the picker is closed', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isKnobIn: false,
                    isPaletteIn: false,
                };
            },
        });

        el.vm.togglePicker();

        expect(el.vm.isKnobIn).toBe(true);
        expect(el.vm.isPaletteIn).toBe(true);
    });

    it('keeps the picker opened when togglePicker is called and the picker is of variant persistent', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
                variant: 'persistent',
            },
            data() {
                return {
                    isKnobIn: true,
                };
            },
        });

        el.vm.togglePicker();

        expect(el.vm.isKnobIn).toBe(true);
    });

    it('stops the press animation & ripple when togglePicker is called', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isRippling: true,
                    isPressed: true,
                };
            },
        });

        el.vm.togglePicker();

        expect(el.vm.isRippling).toBe(false);
        expect(el.vm.isPressed).toBe(false);
    });

    it('hides the palette when hidePalette is called and knob is already hidden', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isKnobIn: false,
                    isPaletteIn: true,
                };
            },
        });

        el.vm.hidePalette();

        expect(el.vm.isPaletteIn).toBe(false);
    });

    it('prevents the palette from hiding when hidePalette is called and knob is not hidden yet', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isKnobIn: true,
                    isPaletteIn: true,
                };
            },
        });

        el.vm.hidePalette();

        expect(el.vm.isPaletteIn).toBe(true);
    });

    it('calls updateColor when onScroll is called and picker is not animating', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.onScroll({
            preventDefault: jest.fn(),
            deltaY: 120,
        });

        expect(updateColorStub).toHaveBeenCalled();
    });

    it('decreases the hue when onScroll is called with an up direction and picker is not animating', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 22,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 22;

        el.vm.onScroll({
            preventDefault: jest.fn(),
            deltaY: -120,
        });

        expect(updateColorStub).toHaveBeenCalledWith(21);
    });

    it('increases hue by 10 when PageUp is pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'PageUp' });

        expect(updateColorStub).toHaveBeenCalledWith(40);
    });

    it('decreases hue by 10 when PageDown is pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'PageDown' });

        expect(updateColorStub).toHaveBeenCalledWith(20);
    });

    it('increases hue by 1 when ArrowUp or ArrowRight are pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'ArrowUp' });
        expect(updateColorStub).toHaveBeenCalledWith(31);

        el.vm.onKeyDown({ key: 'ArrowRight' });
        expect(updateColorStub).toHaveBeenCalledWith(32);
    });

    it('decreases hue by 1 when ArrowDown or ArrowLeft are pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'ArrowDown' });
        expect(updateColorStub).toHaveBeenCalledWith(29);

        el.vm.onKeyDown({ key: 'ArrowLeft' });
        expect(updateColorStub).toHaveBeenCalledWith(28);
    });

    it('sets the hue to 0 when Home is pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'Home' });
        expect(updateColorStub).toHaveBeenCalledWith(0);
    });

    it('sets the hue to 359 when End is pressed', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: false,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.rcp.angle = 30;

        el.vm.onKeyDown({ key: 'End' });
        expect(updateColorStub).toHaveBeenCalledWith(359);
    });

    it('returns early when onKeyDown is called and picker is disabled', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 30,
                disabled: true,
            },
            data() {
                return {
                    isPressed: false,
                    isKnobIn: true,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.onKeyDown({});

        expect(updateColorStub).not.toHaveBeenCalled();
    });

    it('returns early when onScroll is called and picker is animating', () => {
        const updateColorStub = jest.fn();
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
            data() {
                return {
                    isPressed: true,
                    isKnobIn: false,
                };
            },
            methods: {
                updateColor: updateColorStub,
            },
        });

        el.vm.onScroll();

        expect(updateColorStub).not.toHaveBeenCalled();
    });
});

describe('Reactive Changes', () => {
    it('updates `color` based on props', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 90,
                saturation: 100,
                luminosity: 50,
                alpha: 1,
            },
        });

        expect(el.vm.color).toBe('hsla(90, 100%, 50%, 1)');
    });

    it('updates rotator based on hue prop', async () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 90,
                saturation: 100,
                luminosity: 50,
                alpha: 1,
            },
        });

        el.setProps({ hue: 60 });

        await el.vm.$nextTick();

        expect(el.vm.rcp._angle).toBe(60);
    });

    it("doesn't mutate the saturation, luminosity or alpha props when hue changes", () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 90,
                saturation: 100,
                luminosity: 50,
                alpha: 1,
            },
        });

        el.setProps({ hue: 60 });

        expect(el.vm.hue).toBe(60);
        expect(el.vm.saturation).toBe(100);
        expect(el.vm.luminosity).toBe(50);
        expect(el.vm.alpha).toBe(1);
    });

    it('updates `isDragging` to `true` when Rotator starts the drag proccess', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        el.vm.rcp.onDragStart();

        expect(el.vm.isDragging).toBe(true);
    });

    it('updates `isDragging` to `false` when Rotator stops the drag proccess', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        el.vm.rcp.onDragStart();
        el.vm.rcp.onDragStop();

        expect(el.vm.isDragging).toBe(false);
    });
});

describe('Teardown', () => {
    it('cleans up Rotator unused object references & events', () => {
        const el = shallowMount(ColorPicker, {
            propsData: {
                hue: 0,
            },
        });

        el.destroy();

        expect(el.vm.rcp).toBe(null);
    });
});
