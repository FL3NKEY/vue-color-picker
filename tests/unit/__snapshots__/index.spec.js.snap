// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init renders correctly 1`] = `
<div
  aria-disabled="false"
  aria-expanded="true"
  aria-label="color picker"
  aria-roledescription="radial slider"
  aria-valuemax="359"
  aria-valuemin="0"
  aria-valuenow="0"
  aria-valuetext="red"
  class="rcp"
  role="slider"
  tabindex="0"
>
  <div
    class="rcp__palette in"
  >
    <canvas />
  </div>
   
  <div
    class="rcp__rotator"
    style="transform: rotate(0deg); will-change: transform;"
  >
    <div
      class="rcp__knob in"
    />
  </div>
   
  <div
    class="rcp__ripple"
    style="border-color: hsla(0, 100%, 50%, 1);"
  />
   
  <button
    aria-label="color well"
    class="rcp__well"
    style="background-color: hsl(0, 100%, 50%);"
    tabindex="0"
    type="button"
  />
</div>
`;
